core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala

```diff
@@ -41,37 +43,67 @@ class LogDirFailureTest extends IntegrationTestHarness {
   val consumerCount: Int = 1
   val serverCount: Int = 2
   private val topic = "topic"
+  private val partitionNum = 12
 
-  this.logDirCount = 2
+  this.logDirCount = 3
   this.producerConfig.setProperty(ProducerConfig.RETRIES_CONFIG, "0")
   this.producerConfig.setProperty(ProducerConfig.RETRY_BACKOFF_MS_CONFIG, "100")
   this.serverConfig.setProperty(KafkaConfig.ReplicaHighWatermarkCheckpointIntervalMsProp, "60000")
+  this.serverConfig.setProperty(KafkaConfig.NumReplicaFetchersProp, "1")
 
 
   @Before
   override def setUp() {
     super.setUp()
-    TestUtils.createTopic(zkUtils, topic, 1, 2, servers = servers)
+    TestUtils.createTopic(zkUtils, topic, partitionNum, serverCount, servers = servers)
   }
 
   @Test
   def testIOExceptionDuringLogRoll() {
-    testProduceAfterLogDirFailure(Roll)
+    testProduceAfterLogDirFailureOnLeader(Roll)
   }
 
   @Test
   def testIOExceptionDuringCheckpoint() {
-    testProduceAfterLogDirFailure(Checkpoint)
+    testProduceAfterLogDirFailureOnLeader(Checkpoint)
+  }
+
+  @Test
+  def testReplicaFetcherThreadAfterLogDirFailureOnFollower() {
+    val producer = producers.head
```

Comment:
Since we are not producing, do we still need the producer? It seems that we can get the leader/follower info just from the metadata cache in the server?

Reply:
@junrao Sorry for being back-and-forth on this. I realized that the original approach of sending messages with ack = -1 is better than waiting for timestamp to increase. The reason is that, after we mark the partition offline on the follower, we need to wait for follower to fetch from leader and try to append to the local replica, before we verify that the replica fetcher thread is alive.

This will be guaranteed when producer.send() returns. But this is not guaranteed if only check the lastCaughtUpTimeMs -- it is possible that lastCaughtUpTimeMs has increased but replica fetcher thread has not attempted to append to the local replica.

So I reverted the test to use the producer.send() and added an assertion to make sure isr == serverCount. Do you think this approach is reasonable?

Reply:
BTW, could you explain a bit why it is preferred to use metadata cache instead of `producer.partitionsFor()` to get the partition info? It seems that performance is similar. And there is no extra code to construct producer given that producer is already constructed and needed by other test in LogDirFailureTest().